internal/gps/source_test.go

```diff
@@ -37,142 +37,154 @@ func testSourceGateway(t *testing.T) {
-			}
-
-			wantptree := pkgtree.PackageTree{
-				ImportRoot: "github.com/sdboyer/deptest",
-				Packages: map[string]pkgtree.PackageOrErr{
-					"github.com/sdboyer/deptest": {
-						P: pkgtree.Package{
-							ImportPath: "github.com/sdboyer/deptest",
-							Name:       "deptest",
-							Imports:    []string{},
-						},
-					},
-				},
-			}
-
-			ptree, err := sg.listPackages(ctx, ProjectRoot("github.com/sdboyer/deptest"), Revision("ff2948a2ac8f538c4ecd55962e919d1e13e74baf"))
-			if err != nil {
-				t.Fatalf("unexpected err when getting package tree with known rev: %s", err)
-			}
-			if !reflect.DeepEqual(wantptree, ptree) {
-				t.Fatalf("got incorrect PackageTree:\n\t(GOT): %#v\n\t(WNT): %#v", ptree, wantptree)
-			}
-
-			ptree, err = sg.listPackages(ctx, ProjectRoot("github.com/sdboyer/deptest"), NewVersion("v1.0.0"))
-			if err != nil {
-				t.Fatalf("unexpected err when getting package tree with unpaired good version: %s", err)
-			}
-			if !reflect.DeepEqual(wantptree, ptree) {
-				t.Fatalf("got incorrect PackageTree:\n\t(GOT): %#v\n\t(WNT): %#v", ptree, wantptree)
-			}
+	superv := newSupervisor(ctx)
+	sc := newSourceCoordinator(superv, newDeductionCoordinator(superv), cachedir, log.New(test.Writer{TB: t}, "", 0))
+
+	id := mkPI("github.com/sdboyer/deptest")
+	sg, err := sc.getSourceGatewayFor(ctx, id)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if err := sg.require(ctx, sourceIsSetUp); err != nil {
+		t.Fatal(err)
+	}
+
+	wantstate := sourceIsSetUp
+	if sg.srcState != wantstate {
+		t.Fatalf("expected state on initial create to be %v, got %v", wantstate, sg.srcState)
+	}
+
+	if _, ok := sg.src.(*gitSource); !ok {
+		t.Fatalf("Expected a gitSource, got a %T", sg.src)
+	}
+
+	if err := sg.existsUpstream(ctx); err != nil {
+		t.Fatalf("error checking upstream: %s", err)
+	}
+
+	wantstate |= sourceExistsUpstream
+	if sg.src.existsCallsListVersions() {
+		wantstate |= sourceHasLatestVersionList
+		if sg.src.listVersionsRequiresLocal() {
+			wantstate |= sourceExistsLocally & sourceHasLatestLocally
 		}
 	}
+	if sg.srcState != wantstate {
+		t.Fatalf("expected state to be %v, got %v", wantstate, sg.srcState)
+	}
+	if !sg.src.existsCallsListVersions() {
+		_, err := sg.listVersions(ctx)
+		if err != nil {
+			t.Fatalf("Unexpected error getting version pairs from git repo: %s", err)
+		}
+		wantstate |= sourceHasLatestVersionList
+		if sg.src.listVersionsRequiresLocal() {
+			wantstate |= sourceExistsLocally & sourceHasLatestLocally
+		}
+		if sg.srcState != wantstate {
+			t.Fatalf("expected state to be %v, got %v", wantstate, sg.srcState)
+		}
+	}
+
+	if !sg.src.listVersionsRequiresLocal() {
+		if err := sg.syncLocal(ctx); err != nil {
+			t.Fatalf("error on cloning git repo: %s", err)
+		}
 
-	// Run test twice so that we cover both the existing and non-existing case;
-	// only difference in results is the initial setup state.
-	t.Run("empty", do(sourceIsSetUp|sourceExistsUpstream|sourceHasLatestVersionList))
```

Comment:
haven't fully grokked changes to tests here yet, but i do look at the loss of these and worry that we're losing some of the meager set of state-changing cases that we do manage to cover now. no?

Reply:
Looking back at this, it is still valuable to cover existing and non-existing cases (though the states will be the same). I'll restore that. Is there something else that is lost?

Reply:
i don't think so, that's all that i was introducing with the second pass.